presto-main/src/main/java/com/facebook/presto/operator/aggregation/TDigestAggregationFunction.java

```diff
@@ -0,0 +1,206 @@
+
+    private TDigestAggregationFunction(TypeSignature... typeSignatures)
+    {
+        super(
+                NAME,
+                ImmutableList.of(comparableTypeParameter("V")),
+                ImmutableList.of(),
+                parseTypeSignature("tdigest(V)"),
+                ImmutableList.copyOf(typeSignatures));
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Returns a tdigest from the set of doubles";
+    }
+
+    @Override
+    public InternalAggregationFunction specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionManager functionManager)
+    {
+        Type valueType = boundVariables.getTypeVariable("V");
+        TDigestType outputType = (TDigestType) typeManager.getParameterizedType(
+                StandardTypes.TDIGEST,
+                ImmutableList.of(TypeSignatureParameter.of(valueType.getTypeSignature())));
+        return generateAggregation(valueType, outputType, arity);
+    }
+
+    private static InternalAggregationFunction generateAggregation(Type valueType, TDigestType outputType, int arity)
+    {
+        DynamicClassLoader classLoader = new DynamicClassLoader(TDigestAggregationFunction.class.getClassLoader());
+        List<Type> inputTypes = getInputTypes(valueType, arity);
+        TDigestStateSerializer stateSerializer = new TDigestStateSerializer();
+        Type intermediateType = stateSerializer.getSerializedType();
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, outputType.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
+                createInputParameterMetadata(inputTypes),
+                getMethodHandle(valueType, arity),
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION.bindTo(stateSerializer),
+                ImmutableList.of(new AccumulatorStateDescriptor(
+                        TDigestState.class,
+                        stateSerializer,
+                        new TDigestStateFactory())),
+                outputType);
+
+        GenericAccumulatorFactoryBinder factory = AccumulatorCompiler.generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, ImmutableList.of(intermediateType), outputType, true, true, factory);
+    }
+
+    private static List<Type> getInputTypes(Type valueType, int arity)
+    {
+        switch (arity) {
+            case 1:
+                // weight and compression unspecified
+                return ImmutableList.of(valueType);
+            case 2:
+                // weight specified, compression unspecified
+                return ImmutableList.of(valueType, BIGINT);
+            case 3:
+                // weight and compression specified
+                return ImmutableList.of(valueType, BIGINT, DOUBLE);
```

Comment:
Should it be `BIGINT` or some smaller type like `TINYINT`?

Reply:
It could be `TINYINT`, but used `BIGINT` to keep it consistent with `QuantileDigestAggregationFunction`. However, that field represents the weight of the value being added, so it is unlikely that the weight is a `BIGINT`. I do feel more comfortable either keeping the way it is or changing `QuantileDigestAggregationFunction` to `TINYINT` also, but if any user is currently using that function with a `BIGINT` as the weight, it would no longer work, so that might not be the way to go.